### ISSUE TITLE
NullPointerException for has_child query with avg score_mode and explain turned on

### ISSUE THREAD

**Issue by fxh** [elastic/elasticsearch#16525]
Elasticsearch version 2.2.0
java version 1.8.0_25

Steps to reproduce:
```
# clean start
curl -XDELETE localhost:9200/blogs

# prepare parent child relationship
curl -XPUT localhost:9200/blogs -d '{
  "mappings": {
    "blog_tag": {
      "_parent" : {
         "type" : "doc"
      }
    }
  }
}'

# index parent document containing the word 'title'
curl -XPUT 'http://localhost:9200/blogs/doc/1' -d '{
    "message" : "my title"
}'

# index child document NOT containing the word 'title'
curl -XPUT localhost:9200/blogs/blog_tag/1?parent=1 -d '{
    "tag" : "my tag"
}'

# search for any parent or child matching 'title' with explain turned on and
# 'avg' score_mode for child scoring
curl -XGET http://localhost:9200/blogs/_search?pretty -d '{
    "explain": true,
    "query": {
        "bool": {
            "should": [{
                "query_string": {"query": "title"}
            }, {
                "has_child": {
                    "query": {
                        "query_string": {"query": "title"}
                    },
                    "score_mode": "avg",
                    "type": "blog_tag"
                }
            }]
        }
    }
}'
```

The query raises a NullPointerException and returns:

```
[2016-02-09 09:07:13,740][DEBUG][action.search.type       ] [Manta] [blogs][0], node[oYZDW2pHT6SYvh-uTmIKIQ], [P], v[2], s[STARTED], a[id=dnJfwYnXTcahkyT8g019mQ]: Failed to execute [org.elasticsearch.action.search.SearchRequest@611b4d10]
RemoteTransportException[[Manta][127.0.0.1:9300][indices:data/read/search[phase/query+fetch]]]; nested: NullPointerException;
Caused by: java.lang.NullPointerException
	at org.apache.lucene.search.join.GlobalOrdinalsWithScoreCollector$Occurrences.getOccurrence(GlobalOrdinalsWithScoreCollector.java:375)
	at org.apache.lucene.search.join.GlobalOrdinalsWithScoreCollector$Avg.score(GlobalOrdinalsWithScoreCollector.java:229)
	at org.apache.lucene.search.join.GlobalOrdinalsWithScoreQuery$W.explain(GlobalOrdinalsWithScoreQuery.java:131)
	at org.apache.lucene.search.BooleanWeight.explain(BooleanWeight.java:148)
	at org.apache.lucene.search.IndexSearcher.explain(IndexSearcher.java:875)
	at org.apache.lucene.search.IndexSearcher.explain(IndexSearcher.java:852)
	at org.elasticsearch.search.internal.ContextIndexSearcher.explain(ContextIndexSearcher.java:128)
	at org.elasticsearch.search.fetch.explain.ExplainFetchSubPhase.hitExecute(ExplainFetchSubPhase.java:61)
	at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:187)
	at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:478)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryFetchTransportHandler.messageReceived(SearchServiceTransportAction.java:392)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryFetchTransportHandler.messageReceived(SearchServiceTransportAction.java:389)
	at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:350)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[2016-02-09 09:07:13,741][DEBUG][action.search.type       ] [Manta] All shards failed for phase: [query_fetch]
RemoteTransportException[[Manta][127.0.0.1:9300][indices:data/read/search[phase/query+fetch]]]; nested: NullPointerException;
Caused by: java.lang.NullPointerException
	at org.apache.lucene.search.join.GlobalOrdinalsWithScoreCollector$Occurrences.getOccurrence(GlobalOrdinalsWithScoreCollector.java:375)
	at org.apache.lucene.search.join.GlobalOrdinalsWithScoreCollector$Avg.score(GlobalOrdinalsWithScoreCollector.java:229)
	at org.apache.lucene.search.join.GlobalOrdinalsWithScoreQuery$W.explain(GlobalOrdinalsWithScoreQuery.java:131)
	at org.apache.lucene.search.BooleanWeight.explain(BooleanWeight.java:148)
	at org.apache.lucene.search.IndexSearcher.explain(IndexSearcher.java:875)
	at org.apache.lucene.search.IndexSearcher.explain(IndexSearcher.java:852)
	at org.elasticsearch.search.internal.ContextIndexSearcher.explain(ContextIndexSearcher.java:128)
	at org.elasticsearch.search.fetch.explain.ExplainFetchSubPhase.hitExecute(ExplainFetchSubPhase.java:61)
	at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:187)
	at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:478)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryFetchTransportHandler.messageReceived(SearchServiceTransportAction.java:392)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryFetchTransportHandler.messageReceived(SearchServiceTransportAction.java:389)
	at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:350)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[2016-02-09 09:07:13,742][INFO ][rest.suppressed          ] /blogs/_search Params: {pretty=, index=blogs}
Failed to execute phase [query_fetch], all shards failed; shardFailures {[oYZDW2pHT6SYvh-uTmIKIQ][blogs][0]: RemoteTransportException[[Manta][127.0.0.1:9300][indices:data/read/search[phase/query+fetch]]]; nested: NullPointerException; }
	at org.elasticsearch.action.search.type.TransportSearchTypeAction$BaseAsyncAction.onFirstPhaseResult(TransportSearchTypeAction.java:228)
	at org.elasticsearch.action.search.type.TransportSearchTypeAction$BaseAsyncAction$1.onFailure(TransportSearchTypeAction.java:174)
	at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:46)
	at org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:821)
	at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:799)
	at org.elasticsearch.transport.TransportService$4.onFailure(TransportService.java:361)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:39)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: ; nested: NullPointerException;
	at org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:386)
	at org.elasticsearch.action.search.SearchPhaseExecutionException.guessRootCauses(SearchPhaseExecutionException.java:152)
	at org.elasticsearch.action.search.SearchPhaseExecutionException.getCause(SearchPhaseExecutionException.java:99)
	at java.lang.Throwable.printStackTrace(Throwable.java:665)
	at java.lang.Throwable.printStackTrace(Throwable.java:721)
	at org.apache.log4j.DefaultThrowableRenderer.render(DefaultThrowableRenderer.java:60)
	at org.apache.log4j.spi.ThrowableInformation.getThrowableStrRep(ThrowableInformation.java:87)
	at org.apache.log4j.spi.LoggingEvent.getThrowableStrRep(LoggingEvent.java:413)
	at org.apache.log4j.WriterAppender.subAppend(WriterAppender.java:313)
	at org.apache.log4j.WriterAppender.append(WriterAppender.java:162)
	at org.apache.log4j.AppenderSkeleton.doAppend(AppenderSkeleton.java:251)
	at org.apache.log4j.helpers.AppenderAttachableImpl.appendLoopOnAppenders(AppenderAttachableImpl.java:66)
	at org.apache.log4j.Category.callAppenders(Category.java:206)
	at org.apache.log4j.Category.forcedLog(Category.java:391)
	at org.apache.log4j.Category.log(Category.java:856)
	at org.elasticsearch.common.logging.log4j.Log4jESLogger.internalInfo(Log4jESLogger.java:125)
	at org.elasticsearch.common.logging.support.AbstractESLogger.info(AbstractESLogger.java:90)
	at org.elasticsearch.rest.BytesRestResponse.convert(BytesRestResponse.java:131)
	at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:96)
	at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:87)
	at org.elasticsearch.rest.action.support.RestActionListener.onFailure(RestActionListener.java:60)
	at org.elasticsearch.action.search.type.TransportSearchTypeAction$BaseAsyncAction.raiseEarlyFailure(TransportSearchTypeAction.java:316)
	... 10 more
Caused by: java.lang.NullPointerException
	at org.apache.lucene.search.join.GlobalOrdinalsWithScoreCollector$Occurrences.getOccurrence(GlobalOrdinalsWithScoreCollector.java:375)
	at org.apache.lucene.search.join.GlobalOrdinalsWithScoreCollector$Avg.score(GlobalOrdinalsWithScoreCollector.java:229)
	at org.apache.lucene.search.join.GlobalOrdinalsWithScoreQuery$W.explain(GlobalOrdinalsWithScoreQuery.java:131)
	at org.apache.lucene.search.BooleanWeight.explain(BooleanWeight.java:148)
	at org.apache.lucene.search.IndexSearcher.explain(IndexSearcher.java:875)
	at org.apache.lucene.search.IndexSearcher.explain(IndexSearcher.java:852)
	at org.elasticsearch.search.internal.ContextIndexSearcher.explain(ContextIndexSearcher.java:128)
	at org.elasticsearch.search.fetch.explain.ExplainFetchSubPhase.hitExecute(ExplainFetchSubPhase.java:61)
	at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:187)
	at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:478)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryFetchTransportHandler.messageReceived(SearchServiceTransportAction.java:392)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryFetchTransportHandler.messageReceived(SearchServiceTransportAction.java:389)
	at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:350)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	... 3 more
{
  "error" : {
    "root_cause" : [ {
      "type" : "null_pointer_exception",
      "reason" : null
    } ],
    "type" : "search_phase_execution_exception",
    "reason" : "all shards failed",
    "phase" : "query_fetch",
    "grouped" : true,
    "failed_shards" : [ {
      "shard" : 0,
      "index" : "blogs",
      "node" : "oYZDW2pHT6SYvh-uTmIKIQ",
      "reason" : {
        "type" : "null_pointer_exception",
        "reason" : null
      }
    } ]
  },
  "status" : 500
}
```

'avg' is the only score_mode that raises this exception, other modes work correctly. Also if I turn explain off, the exception is not raised.
In version 1.3.7 of Elasticsearch the above sequence works without raising an exception. I have not tested any version in-between.

**Comment by martijnvg** [comment]
@fxh Thanks for reporting, this is clearly a bug.

**Comment by martijnvg** [comment]
This error is caused by a bug in the joining code in Lucene. The joining code always tries to compute a score explanation even in the case that a document doesn't match. With `score_mode` set to `avg` this causes a NPE and with other score modes it returns a score explanation that indicates that the document matched while it didn't.

I opened an issue in Lucene to fix the problem: https://issues.apache.org/jira/browse/LUCENE-7065

**Comment by clintongormley** [comment]
[LUCENE-7065](https://issues.apache.org/jira/browse/LUCENE-7065) has been fixed in Lucene 6.0, and so will be fixed in Elasticsearch 5.0.0.
